### PR TITLE
fix devcontainer to be ready for development

### PR DESCRIPTION
**Pull request by serbrech**
# DevContainer

The current DevContainer is outdated and when starting a codespace, you need to install a newer node, npm, typespec, etc...

This PR changes the devcontainer configuration so that the codespace is properly configured to start editing the specs.

- use `mcr.microsoft.com/devcontainers/typescript-node:20-bookworm` as the base image
- install `packages-microsoft-prod.deb`
- ~~install git~~
- install npm oav
 
 
 on environment start:
- install npm packages from packages.json (typespec compiler amongst other)
- set the `node_modules/.bin` folder to the path, so we can call `tsp`

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -3,10 +3,28 @@
...
+    "ghcr.io/devcontainers/features/powershell:1": {
+      "version": "latest"
+    }
+  },
+  "customizations": {
+    // Configure properties specific to VS Code.
+    "vscode": {
+      // Add the IDs of extensions you want installed when the container is created.
+      "extensions": [
+        "esbenp.prettier-vscode"
```
Comment (timotheeguerin) [inline .devcontainer/devcontainer.json:26]: Typespec extension too?

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -3,10 +3,28 @@
...
-  "mounts": [
-    "source=/home/rc/repos/tmp/sdk-repos,target=/workspaces/sdk-repos,type=bind,consistency=cached"
-  ]*/
+  "postCreateCommand": "npm install",
+  "remoteUser": "root",
+  "remoteEnv": { 
+    "PATH": "${containerEnv:PATH}:${containerWorkspaceFolder}/node_modules/.bin" 
+  },
+  "features": {
+    "ghcr.io/devcontainers/features/node:1": {
```
Comment (timotheeguerin) [inline .devcontainer/devcontainer.json:12]: More of a question but what does this do on top of already using the node base image?

Comment (serbrech) [reply .devcontainer/devcontainer.json:12]: Very good point :)
it does nothin I'll remove it.
I started with a different image, where this was necessary, then moved to the node-typescript base image, which uses node 20 already:

➜  ~ docker run -it --rm mcr.microsoft.com/devcontainers/typescript-node:20-bookworm /bin/bash
root ➜ / $ node --version
v20.13.1
root ➜ / $ npm --version
10.5.2

removing int

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (serbrech) [reply .devcontainer/devcontainer.json:26]: sure

Comment (serbrech) [reply .devcontainer/devcontainer.json]: done

File: .devcontainer/Dockerfile [hunk 3]
```diff
@@ -1,21 +1,13 @@
...
 
-RUN apt-get -y update &&  apt upgrade -y && apt install curl -y
-RUN apt-get -y install git
+WORKDIR /workspaces/azure-rest-api-specs
 
-# install node and npm
-RUN curl -sL https://deb.nodesource.com/setup_16.x | bash -
-RUN apt -y install nodejs
-RUN npm install -g npm@9.5.0
+RUN wget -q https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb \
```
Comment (timotheeguerin) [inline .devcontainer/Dockerfile:5]: Do you also know what this one does?

Comment (serbrech) [reply .devcontainer/Dockerfile:5]: yes, it configures the microsoft packages sources:

https://learn.microsoft.com/en-us/linux/packages#how-to-install-microsoft-software-packages-using-the-linux-repository

Comment (mikeharder) [reply .devcontainer/Dockerfile:5]: I think we should be able to simplify the Dockerfile even further.  It should just need:

1. Node 20
2. clone repo
3. npm ci

File: .devcontainer/Dockerfile
Comment (mikeharder) [inline .devcontainer/Dockerfile:1]: I think we can delete this file.  We should be able to use a standard image.

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -3,10 +3,24 @@
   "build": {
     "dockerfile": "Dockerfile"
```
Comment (mikeharder) [inline .devcontainer/devcontainer.json:4]:
```suggestion
    "image": "mcr.microsoft.com/devcontainers/typescript-node:dev-20"
```

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -3,10 +3,24 @@
...
   },
-  "remoteUser": "root"
-  /*,
-  update below source folder to your local sdk-repos folder
-  "mounts": [
-    "source=/home/rc/repos/tmp/sdk-repos,target=/workspaces/sdk-repos,type=bind,consistency=cached"
-  ]*/
+  "postCreateCommand": "npm install",
```
Comment (mikeharder) [inline .devcontainer/devcontainer.json:6]:
```suggestion
  "postCreateCommand": "npm ci",
```

File: .devcontainer/devcontainer.json [hunk 6]
```diff
@@ -3,10 +3,24 @@
...
+  "remoteUser": "root",
+  "remoteEnv": { 
+    "PATH": "${containerEnv:PATH}:${containerWorkspaceFolder}/node_modules/.bin" 
+  },
+  "features": {
+    "ghcr.io/devcontainers/features/powershell:1": {
+      "version": "latest"
+    }
+  },
```
Comment (mikeharder) [inline .devcontainer/devcontainer.json:15]:
```suggestion
```

File: .devcontainer/devcontainer.json [hunk 7]
```diff
@@ -1,12 +1,20 @@
 {
   "name": "TypeSpec Dev Container",
-  "build": {
-    "dockerfile": "Dockerfile"
+  "image": "mcr.microsoft.com/devcontainers/typescript-node:dev-20",
+  "postCreateCommand": "npm ci",
+  "remoteEnv": { 
+    "PATH": "${containerEnv:PATH}:${containerWorkspaceFolder}/node_modules/.bin" 
   },
```
Comment (mikeharder) [inline .devcontainer/devcontainer.json:7]:
```suggestion
```

I think you should be able to remove this, it should just work by default.

File: .devcontainer/devcontainer.json [hunk 8]
```diff
@@ -1,12 +1,20 @@
...
+      // Add the IDs of extensions you want installed when the container is created.
+      "extensions": [
+        "EditorConfig.EditorConfig",
+        "esbenp.prettier-vscode",
+        "typespec.typespec-vscode",
+        "docsmsft.docs-authoring-pack"
+      ]
+    }
+  }
 }
```
Comment (mikeharder) [inline .devcontainer/devcontainer.json:20]:
```suggestion
}

```

File: .devcontainer/devcontainer.json [hunk 9]
```diff
@@ -1,12 +1,17 @@
 {
   "name": "TypeSpec Dev Container",
```
Comment (mikeharder) [inline .devcontainer/devcontainer.json:2]:
```suggestion
  "name": "azure-rest-api-specs Dev Container",
```

File: .devcontainer/devcontainer.json [hunk 10]
```diff
@@ -1,12 +1,17 @@
...
-  },
-  "remoteUser": "root"
-  /*,
-  update below source folder to your local sdk-repos folder
-  "mounts": [
-    "source=/home/rc/repos/tmp/sdk-repos,target=/workspaces/sdk-repos,type=bind,consistency=cached"
-  ]*/
-}
\ No newline at end of file
+  "name": "azure-rest-api-specs Dev Container",
```
Comment (mikeharder) [inline .devcontainer/devcontainer.json:2]:
```suggestion
  "name": "TypeSpec Dev Container",
```

**Review by mikeharder — APPROVED** [review]

**Comment by mikeharder** [comment]
> set the node_modules/.bin folder to the path, so we can call tsp

Better to get into the habit of running `npx tsp`, or just install the matching version of `tsp` globally.

**Pull request merged**